### DISCUSSION TITLE
Using arduino uno instead of mega.

### DISCUSSION THREAD

**Discussion by kilehh** [tttapa/Control-Surface#559] [Q&A] [answered]
Hello!
I have an diy midi device which has 15 potentiometers connected using multiplexer (CD74HC4067), 40 buttons using button matrix, 7 buttons and 3 encoders are just connected to the arduino mega 2560, and i am planing to add a quite a bit of leds using shift registers. The problem is that the arduino mega has the CH340, and i have to use Hairless MIDI Serial Bridge in order to use my device on pc, but since i already have a arduino uno which has a 16U2, i was thinking about moving all that from mega to uno. The question is, would that be somehow possible, and if yes how (maybe with help of stacking multiplexers etc), or maybe there is any other solution that would let me use the device without programs like Hairless MIDI Serial Bridge.
Thanks in advance!

**Comment by tttapa** [comment, marked as answer]
It depends on how much memory your code uses. The Uno only has 2KiB of RAM, which is very limited.  
The Uno only has 2 interrupt pins, so you'll have to jump through some hoops to get 3 encoders working, e.g. using pin change interrupts or timer interrupts, I don't think polling is going to be fast enough, given the number of analog inputs you have to read.

**Reply by kilehh** [reply]
That means buying mega with 16U2 is the only option?

**Reply by tttapa** [reply]
No, there are many alternatives, see https://tttapa.github.io/Control-Surface-doc/Doxygen/d8/d4a/md_pages_MIDI-over-USB.html.

In my opinion, the Arduino Uno and Mega are not suitable for building MIDI controllers. They're slow, have limited memory, do not support MIDI over USB, have a slow ADC, a limited number of interrupt-capable pins for encoders, etc.
